polaris-patterns/src/components/PolarisPatternsProvider/PolarisPatternsProvider.tsx

```diff
@@ -0,0 +1,60 @@
+import type {PropsWithChildren} from 'react';
+import React, {useEffect} from 'react';
+import {I18nextProvider} from 'react-i18next';
+
+import {DEFAULT_I18N_DETAILS, I18NEXT_NAMESPACE} from './constants';
+import {i18next, I18nDetailsContext} from './i18n';
+
+export interface TranslationDictionary {
+  [key: string]: string | TranslationDictionary;
+}
+
+export interface I18nDetails {
+  locale: string;
+  region?: string;
+  currencyCode?: string;
+  timeZone?: string;
+}
+
+export interface Props {
+  /** User's current locale information.
+   * @default {locale: 'en', country: 'US', currency: 'USD'}
+   */
+  i18nDetails?: I18nDetails;
+  /** A locale's dictionary object. If not provided, only default english translations will be loaded. */
+  translations?: TranslationDictionary;
+}
+
+export function PolarisPatternsProvider({
+  translations,
+  i18nDetails = DEFAULT_I18N_DETAILS,
+  children,
+}: PropsWithChildren<Props>) {
+  useEffect(() => {
+    if (!i18next.hasResourceBundle(i18nDetails.locale, I18NEXT_NAMESPACE)) {
+      i18next.addResourceBundle(
+        i18nDetails.locale,
+        I18NEXT_NAMESPACE,
+        translations,
+      );
+    } else if (i18nDetails.locale === DEFAULT_I18N_DETAILS.locale) {
+      i18next.addResourceBundle(
+        i18nDetails.locale,
+        I18NEXT_NAMESPACE,
+        translations,
+        true,
+        true,
+      );
+    }
```

Comment:
If the bundle for a certain locale has already been added, I don't overwrite it. It would be kind of odd behaviour for a user to provide two different dictionaries for the same locale at two different times.

I do have that second case where if the locale is english, I overwrite the existing `en` resource. I do this because the en resource is loaded on `init` and if the user passes in custom translations, it would skip the first conditional when we want to overwrite the default `en.json`

Unfortunately this will rewrite the `en` translations everytime a user changes their language to english but users shouldn't be switching locales frequently